rtsp.go

```diff
@@ -61,10 +62,38 @@ var (
 	ErrH264PassthroughNotEnabled = errors.New("H264 passthrough is not enabled")
 )
 
+func getStringFromExtra(extra map[string]interface{}, key string) (string, error) {
+	extraVal, ok := extra[key]
+	if !ok {
+		return "", fmt.Errorf("key not found in 'extra': %s", key)
+	}
+	extraValStr, ok := extraVal.(string)
+	if !ok {
+		return "", errors.New("'extra' value must be a string")
+	}
+	return extraValStr, nil
+}
+
 func init() {
 	for _, model := range Models {
 		resource.RegisterComponent(camera.API, model, resource.Registration[camera.Camera, *Config]{
 			Constructor: newRTSPCamera,
+			Discover: func(_ context.Context, logger logging.Logger, extra map[string]interface{}) (interface{}, error) {
+				logger.Debugf("viamrtsp discovery received extra credentials: %v", extra)
+				username, err := getStringFromExtra(extra, "username")
+				if err != nil {
+					return nil, err
+				}
+				password, err := getStringFromExtra(extra, "password")
+				if err != nil {
+					return nil, err
+				}
+				camInfoList, err := viamonvif.DiscoverCameras(username, password, logger)
+				if err != nil {
+					return nil, err
+				}
```

Comment:
should we log and still create if discover fails? Or will it necessarily mean that rtsp cameras are not on network and can't be reached?

Reply:
Also, will this work for all models in this constructor?

Reply:
> Also, will this work for all models in this constructor?
I think so since it's in the for each loop over all models registering

> should we log and still create if discover fails? Or will it necessarily mean that rtsp cameras are not on network and can't be reached?
Taking a look at `DiscoverCameras`, there are two returns
`return nil, fmt.Errorf("failed to retrieve network interfaces: %w", err)` and `return &CameraInfoList{Cameras: discoveredCameras}, nil`

So no error is blocking except for when we cannot discover the available network interfaces to call ws-discovery on, which feels like a big enough issue to just fail on I think

Reply:
okay, it's an easy change if it is catastrophic for someone who knows their rtsp camera ids anyway, so we can change that behaviour with more feedback.

Reply:
just to double check, the Discover function makes sense for the other models in this module, the `mjpeg`, the `h265` etc.

Reply:
> just to double check, the Discover function makes sense for the other models in this module, the mjpeg, the h265 etc.

I think it does yeah. All the models we support require an rtsp address

Reply:
> If Discovery errors, does RegisterComponent also fail?

No since its just being registered as a callback not being called